### DISCUSSION TITLE
Match Type is missing in DynamicLinks

### DISCUSSION THREAD

**Discussion by ex3ndr** [invertase/react-native-firebase#5583] [General]
Match Type is critical for measuring performance of Dynamic Links and yet it absent from JS api.

**Comment by mikehardy** [comment]
It appears to not exist on firebase-android-sdk (please if you can find it in their reference documentation paste a link here - happy to be corrected).

It does appear to exist on iOS https://firebase.google.com/docs/reference/ios/firebasedynamiclinks/api/reference/Classes/FIRDynamicLink#matchtype

If you would like to post a PR that adds it to index.d.ts (marked as iOS only, assuming I am correct that it does not exist on android) and fetch it, I'd be happy to collaborate with you and get it published.

Add a new enum with none / weak / default / unique string values to index.d.ts and add a new matchType property to the link object that references that enum type here - https://github.com/invertase/react-native-firebase/blob/d4b2015f8def4759b95072cd4bca86eda0443c54/packages/dynamic-links/lib/index.d.ts#L399-L424

Two similar blocks, both would fetch the match type and set a new key/value in the returned object, where the key was as indicated in the index.d.ts addition and the value was one of the enumeration values possible per the index.d.ts addition

The code doing the transformation from resolved link to JSON object for transmission as an event to the javascript is unfortunately really duplicated, but it is also just the same two lines over and over again, e.g.:

https://github.com/invertase/react-native-firebase/blob/d4b2015f8def4759b95072cd4bca86eda0443c54/packages/dynamic-links/ios/RNFBDynamicLinks/RNFBDynamicLinksAppDelegateInterceptor.m#L68-L71

https://github.com/invertase/react-native-firebase/blob/d4b2015f8def4759b95072cd4bca86eda0443c54/packages/dynamic-links/ios/RNFBDynamicLinks/RNFBDynamicLinksAppDelegateInterceptor.m#L89-L92

...and in the other implementation file:

https://github.com/invertase/react-native-firebase/blob/d4b2015f8def4759b95072cd4bca86eda0443c54/packages/dynamic-links/ios/RNFBDynamicLinks/RNFBDynamicLinksModule.m#L129-L137 (plus a few other areas in the file)

**Comment by ingstartup** [comment]
matchType is not needed for Android b/c there's platform support for deferred deep link. They are always 100% confident about the trigger of a deeplink.

On the other hand, iOS has no platform support on deferred deep link. There are third party solution (in this case Firebase Dynamic Link) trying to place a hack around it under the hood, and hence the matchType is needed on iOS. it tells the confidence of whether or not the deep link was actually clicked.

Right now, the bug that I'm seeing is that if I open a dynamic link and the app is not installed, then it will take me to the appstore and then after the installation, it takes me to the right screen. This is great so far. However, the dynamic link is not consumed! 
If I uninstall the app and install it again, even without clicking on the link again, a deep link is still triggered and take me to the specified screen.
See a community post here on StackOverFlow: https://stackoverflow.com/questions/47908554/once-opening-a-firebase-link-always-returns-a-link-when-reinstalling-the-app-on/47998881#47998881. It was mentioned that the link actually lingers for an hour before it expires.

The hope is that with the matchType, we can place some logic to help filter the low confidence "deep link" and mitigate this bug.

**Reply by mikehardy** [reply]
Very interesting. I noticed the next comment mentioned you should add the domains to Info.plist - I myself have mine in Info.plist and I have not noticed a failure to consume links ? That said, my comment above was pretty prescriptive - I don't have the time to do this and test it myself right now but I would love to see the feature and would happily collaborate with anyone to enable it

**Reply by ingstartup** [reply]
According to the doc, https://rnfirebase.io/dynamic-links/usage#dynamic-links-with-custom-domains, this is only required if you're using custom domain. i'm just using the default xxxxx.page.link

I'm not sure if this is related unless it's some undocumented behavior?

**Reply by mikehardy** [reply]
It seems to only rhyme at this point - maybe related, but no conclusive proof, I lean towards undocumented behavior at the moment, with matchType being a reasonable workaround on iOS for non-custom-domains

**Reply by ingstartup** [reply]
I just gave it a shot. It doesn't help. The dynamic link is still remembered in subsequent re-installation cycle even we only have clicked it once before the first installation.

And i've also re-confirmed that it took ~1 hour before that state is expired.

**Reply by ingstartup** [reply]
The matchType is documented here https://firebase.google.com/docs/reference/swift/firebasedynamiclinks/api/reference/Classes/DynamicLink#matchtype

it's a property of dyanmic link, very similar to minimumAppVersion and  url.
I think we just need to add
```
@"matchType": dynamicLink.matchType
``` 
to the places you mentioned above at `RNFBDynamicLinksAppDelegateInterceptor.m` and 
`RNFBDynamicLinksModule.m`
